### ISSUE TITLE
[css-values-4][css-writing-modes-4] Revisit decision to use 永 instead of 水 as the ic unit

### ISSUE THREAD

**Issue by ziyunfei** [w3c/csswg-drafts#7577]
Chrome just implemented the ic unit this week, when I shared this news in Chinese social media, some people are confused about the chosen character 水(water), especially those people who had the handwriting practice experience, they think 永(forever; eternal) is the right choice.

See this Twitter thread: https://twitter.com/intenttoship/status/1555274307735285760

**Comment by tiye** [comment]
quote https://en.wikipedia.org/wiki/Eight_Principles_of_Yong

> It was traditionally believed that the frequent practice of these principles as a beginning [calligrapher](https://en.wikipedia.org/wiki/Chinese_calligraphy) could ensure beauty in one's writing.

**Comment by yisibl** [comment]
See https://github.com/w3c/csswg-drafts/issues/2798

**Comment by yisibl** [comment]
If CSSWG agrees to this change, I can amend the relevant specification and WPT.

**Comment by frivoal** [comment]
> See #2798

In particular, this comment https://github.com/w3c/csswg-drafts/issues/2798#issuecomment-398668398 says:
> [Writing Modes uses 水 as well](https://www.w3.org/TR/2018/CR-css-writing-modes-3-20180524/#text-combine-compression) (used when scaling down the tate-chu-yoko to 1em). If switching to 永, I kinda want to see two specs refer to the same character.

**Comment by tabatkins** [comment]
And note also in #2798 that the reason we stuck with 水 rather than switching to 永 is solely because it would have been annoying to update Writing Modes, and in practice the two should be roughly identical in font coverage anyway.

But if people keep getting confused about it, it's probably worth going thru the effort of switching the character.

**Comment by ziyunfei** [comment]
Filed bugs:
https://bugzilla.mozilla.org/show_bug.cgi?id=1783937
https://crbug.com/1351509
https://bugs.webkit.org/show_bug.cgi?id=243746

**Comment by VaJoy** [comment]
Using `永` is the better one cuz it is a standard of Asian calligraphy.

I promise almost every devloper in China, Japan, Korea would be confused while using `水` as the unit of `ic` so please switch to `永` to avoid this disaster.

**Comment by VaJoy** [comment]
Using `永` is the better choice cuz it is a standard of Asian calligraphy.

I promise almost every devloper in China, Japan, Korea would be confused while using `水` as the unit of `ic` so please switch to `永` to avoid this disaster.

**Comment by gongpeione** [comment]
IMO using `永` as the ic unit makes more sense in Chinese and also looks more professional.

**Comment by myakura** [comment]
Although I'm not strongly against about it, I don't think it's worth the change to the specs and implementations.

I don't know how things are going on in #3135 , but if the `ic` requires implementations to download the font, then 永 (U+6C38) has higher chance to cause extra font download compared to 水 (U+6C34):

* In [Noto Sans Japanese](https://fonts.googleapis.com/css2?family=Noto+Sans+JP), 水 (U+6C34) is placed in the 5th most used characters bucket, whereas 永 (U+6C38) is in the 22nd
* In [Noto Sans Traditional Chinese](https://fonts.googleapis.com/css2?family=Noto+Sans+TC), 水 is in the 2nd bucket, and 永 is in the 5th
* In [Noto Sans Simplified Chinese](https://fonts.googleapis.com/css2?family=Noto+Sans+SC), 水 is in the 3rd bucket, and 永 is in the 6th
* In [Noto Sans Hong Kong](https://fonts.googleapis.com/css2?family=Noto+Sans+TC), 水 is in the 2nd bucket, and 永 is in the 5th
* In [Noto Sans Korean](https://fonts.googleapis.com/css2?family=Noto+Sans+KR), 水 is in the 16th bucket, and 永 is in the 39th

**Comment by Naeemo** [comment]
IMO, using `永` would be a more natural choice.
I would be confused by `水` and try to find out why, but not with `永`. `永` is the first thing to practice for properly Chinese handwriting cuz its structure, it's kind of common sense here.

**Comment by yisibl** [comment]
In fact, the `永` is used in CSS Inline L3. It should also be modified to be consistent from the point of view of consistency of the specification.

> The bounding box of 永 (U+6C38) can be used to find the ideographic character face edges.
> https://drafts.csswg.org/css-inline-3/#baseline-synthesis-fonts

**Comment by foolip** [comment]
Is the issue here the belief that 水 was picked by accident because someone conflated it with 永?

@fantasai disconfirmed that in https://twitter.com/fantasai/status/1555385335663759360, and I wonder if the same issue would be raised if the chosen character was instead 口 or something.

I'm not a member of the CSSWG, but the fact that 永 is a kind of reference character in calligraphy doesn't by itself make it a good choice for the `ic` unit. The kind of analysis that @myakura did in https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1210082393 seems like a better ground for this kind of decision.

**Comment by yisibl** [comment]
@myakura @foolip 

I'm sorry, but I don't endorse the reasons related to font slicing in Google Web Fonts.

The same problem would exist in the ch unit, which should be proposed to Google Fonts to change to put the characters that are CSS units (`0` or `永`) further ahead.

I think #3135 about downloading fonts is a separate issue, not related to the proposal of this issue.

**Comment by bnoctis** [comment]
A rather less technical question: Who decides how Chinese characters should be rendered? People who speak and understand Chinese, or those who don't?

**Comment by tiye** [comment]
@bnoctis to be accurate, this character is also used by some other Asian countries although they do speak Chinese in daily life.

**Comment by bnoctis** [comment]
@tiye To be accurate, CJK characters, which still holds, for people using CJK languages.

**Comment by azbo** [comment]
Support rewriting to "永"

**Comment by sout233** [comment]
“永”is a better choice

**Comment by cxh194311** [comment]
Support rewriting to "永"

**Comment by woeoio** [comment]
Support rewriting to "永",

**Comment by vicdf** [comment]
As a chinese, I think 永 is way more representative compare to 水 due to its structure, etc. so I hope it is the best to change that.

**Comment by dufemeng** [comment]
I agree with the above statement,  support rewriting to "永"

**Comment by aaron-ai** [comment]
"永" is better.

**Comment by wlbksy** [comment]
When used in calligraphy, "水" does not make any sense. "永" is the right choice for it represent the basic skills of calligraphy, "Eight Principles of Yong" https://en.wikipedia.org/wiki/Eight_Principles_of_Yong.
Almost every calligrapher knows this as beginner. 
It is the  essential sound of "mama/papa" to babies who are learning how to speak.

**Comment by hingbong** [comment]
I think "永" is better, too.

**Comment by linrf** [comment]
I think "永" is better.

**Comment by Rvtea** [comment]
"永" is better.

**Comment by volcanicll** [comment]
Support rewriting to "永"

**Comment by ziyunfei** [comment]
Guys, please stop posting similar comments unless you have more informed thoughts. you can click the 👍🏻 button to vote on it.

**Comment by DreamOneX** [comment]
As a Chinese who has studied calligraphy, I know `Eight Principles of Yong` has extraordinary meaning. It is the basic skill of calligraphy. I believe every developer from China, Japan and Korea will support using `永` instead of `水`. 
If you know Chinese, even if you haven't learned calligraphy, You can also find that the character `永` contains all eight Chinese strokes. So it can make more sense.
What's more, as @bnoctis said, after all we use Chinese, our culture should be respected and we should also decide how to render our characters.
All in all, I support rewriting to `永`.
thanks.

**Comment by bobwxc** [comment]
In most cases, 永  is the first example character of traditional ShuFa Art. And it also be the simplest example character of much fonts. Like 'n' in many Latin fonts.
It is good traditional customs.

**Comment by y122972** [comment]
`永` is better.

**Comment by weolar** [comment]
"永" is better. it is most familiar character of traditional Chinese Calligraphy Art. Support rewriting to "永"

**Comment by sword4869** [comment]
i agree to use 永, which is a great idea to adapt the font design for the culture behind the font.

**Comment by ArthurMao** [comment]
The Eight Principles of Yong (Chinese: 永字八法; pinyin: Yǒngzì Bā Fǎ; Japanese: 永字八法/えいじはっぽう, eiji happō; Korean: 영자팔법/永字八法, Yeongjapalbeop; Vietnamese: Vĩnh tự bát pháp) explain how to write eight common strokes in regular script which are found all in the one character, 永 (pinyin: yǒng, "forever", "permanence"). It was traditionally believed that the frequent practice of these principles as a beginning calligrapher could ensure beauty in one's writing.

**Comment by Yyang888** [comment]
ahhaha鸡汤来历

**Comment by yuan3721** [comment]
Support rewriting to "永"

**Comment by Yyang888** [comment]
这是来自邮箱的自动回复邮件。暑假内卷学习中www 看到会及时回的！

**Comment by BI1LQV** [comment]
"永" is better for its meaning

**Comment by 1219521375** [comment]
As a Chinese,I think "永" is better.

**Comment by SkyeYoung** [comment]
As a Chinese, I think '永' is better.

Take an analogy that I'm not too sure is appropriate: the letter 'x' is used as the x-height in typography, but in fact 'v', 'w' and 'z' are similar in height, so why not say v-height, w-height? Would it be normal for people with English as their native language if a person said v-height or w-height?

> I know there is a claim that 'x' is really the only letter in all serif fonts that touches both lines. But the statement above comes from https://en.wikipedia.org/wiki/X-height.

In any case, at least for native speakers of CJK, the use of '水' instead of '永' would make them feel very strange.

**Comment by Siricee** [comment]
> Although I'm not strongly against about it, I don't think it's worth the change to the specs and implementations.
> 
> [#7577 (comment)](https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1209711033)
> 
> > and in practice the two should be roughly identical in font coverage anyway.
> 
> Speaking of font coverage, they are not the same in _web_ fonts. I don't know how things are going on in #3135 , but if `ic` requires implementations to download the font, then 永 (U+6C38) has higher chance to cause extra font download compared to 水 (U+6C34):
> 
> * In [Noto Sans Japanese](https://fonts.googleapis.com/css2?family=Noto+Sans+JP), 水 (U+6C34) is placed in the 5th most used characters bucket, whereas 永 (U+6C38) is in the 22nd
> * In [Noto Sans Traditional Chinese](https://fonts.googleapis.com/css2?family=Noto+Sans+TC), 水 is in the 2nd bucket, and 永 is in the 5th
> * In [Noto Sans Simplified Chinese](https://fonts.googleapis.com/css2?family=Noto+Sans+SC), 水 is in the 3rd bucket, and 永 is in the 6th
> * In [Noto Sans Hong Kong](https://fonts.googleapis.com/css2?family=Noto+Sans+TC), 水 is in the 2nd bucket, and 永 is in the 5th
> * In [Noto Sans Korean](https://fonts.googleapis.com/css2?family=Noto+Sans+KR), 水 is in the 16th bucket, and 永 is in the 39th

As a Chinese, the priority of bucket sorting is lower than the priority of meaning in culture. '永' contains all Chinese language strokes, rather than '水'. In Typography Design , '永' is always used most common.

Here are some references:

[Designing with the Hanzi script – Keith Tam](https://keithtam.net/designing-with-the-hanzi-script/)

![pic](https://keithtam.net/wp-content/uploads/2019/08/Fig-1-1-195x300.png)

[Can a Chinese font be analysed and criticised like a Western (alphabetical) one? - Quora](https://www.quora.com/Can-a-Chinese-font-be-analysed-and-criticised-like-a-Western-alphabetical-one)

> A good way to understand the varieties of strokes is by looking at the character 永 (yǒng), which contains the 8 most common strokes in the regular script. Together, they are known as the “Eight Principles of Yong.”
>
> ![pic](https://qph.cf2.quoracdn.net/main-qimg-0616fc75009ad69c48563227288bc247)
> 
> Although the Eight Principles are a calligraphy concept, the terms for which are shared with typography.
> ![pic2](https://qph.cf2.quoracdn.net/main-qimg-49c2094f8acbad1fc6657e470e114d0b-pjlq)

Lastly, the use of '水' or '永', the choice should be made by at least a native speakers of CJK.

**Comment by sparklerl** [comment]
As a Chinese,I think "永" is better .

**Comment by Matthewow** [comment]
support

**Comment by sunhaitao** [comment]
I'm afraid that we need do some extra works to ensure '永' is as functional as '水' first.

the '㇔' in '永' is a stroke can be drawn in various forms to make a style. For example, it can be drawn like a '㇐' to create a more formal, historic, stable feeling (See https://dribbble.com/shots/4558601-/attachments/10464590?mode=media). 

Since the 'ic' unit represents an advance measure, and an advance measure can be both width or height of a glyph (depending on the writing direction), whether '永' in various fonts always have representative heights is needed to be verified ('水' may be a safer bet here for its simpler structure).

**Comment by Baoge2012** [comment]
"永" is a much more Chinese way

**Comment by bnoctis** [comment]
@sunhaitao The character may change drastically in different typefaces, but if every typeface follows the aforementioned principle, 永 would still serve as the basis, holding the position. Unless `ic` is a cross-typeface unit, in which case I suggest 一 as it's a single horizontal stroke with no other parts and fits as a width measuring unit.

**Comment by sunhaitao** [comment]
> but if every typeface follows the aforementioned principle

I'm afraid this cannot be assumed for the following reasons:

1. In the history of computing, it is rare that many vendors all follow a specification strictly. People (and maybe all beings with general intelligence) just made mistakes here and there. 

2. Although translated as 'Eight Principles of Yong' on Wikipedia, '永字八法' is actually 8 mostly-used stroke-drawing techniques demonstrated with the 'forever' glyph. And it is mostly for the regular script (楷体). More ancient scripts such as the Clerical script (隶书) have their own rules. Of course, many (if not most) typefaces used today are based on the regular script for readability. But artistic fonts are different stories.

'永' is the good old choice to show strokes. But representing measurements is a new problem, picking another character is also okay. Since '水' is neither found dysfunctional nor associated with something unjust or distasteful, we are not in a hurry to abandon the current work. Let's resolve this cautiously.

**Comment by bnoctis** [comment]
@sunhaitao I agree on your technical ground. However, I insist that such standards be discussed and formalized by native speakers of the language, not some random Google developer speaking another.

**Comment by faceless2** [comment]
Please remember we are exclusively discussing the **OpenType glyph advance** here, not any other aspect of font metrics, not calligraphy or aesthetics. Specifically the values for these glyphs from the OpenType `hmtx` and `vmtx` tables.

In practice the advance will be the same for both glyphs, in both the horizontal and vertical directions.

**Comment by CoelacanthusHex** [comment]
> In practice the advance will be the same for both glyphs, in both the horizontal and vertical directions.

Since they are the same in practice, why not choose the one that respects the tradition of the language more.

**Comment by bnoctis** [comment]
> Please remember we are exclusively discussing the **OpenType glyph advance** here, not any other aspect of font metrics, not calligraphy or aesthetics. Specifically the values for these glyphs from the OpenType `hmtx` and `vmtx` tables.

I suppose glyph advance is based on glyphs, which are the forming factor of a typeface, which respects traditions because it evolved with them.

This is a too professional topic requiring experts to decide on technical details, but I wonder how well a non-native user of a language can do.

**Comment by sunhaitao** [comment]
@bnoctis
According to my understanding of the Chinese culture, 'where the heroes are from' does not matter' (英雄不问出处). Anyone with enough domain knowledge is eligible for making design decisions. In this particular problem, a typography expert studying Chinese as a foreign language may have more domain knowledge. Cultures have ambiguous parts that only native users of associated languages can handle well. But measuring a glyph is not that hard.

On the other hand, foreign users are also stakeholders in layouting characters. People do publish materials in their foreign languages (this page is an example). And mixing native and foreign characters in the same sentence is a valid use case (this page is also an example). Problems in such practices (some are hard to imagine for native users) also need to be addressed.

By the way, as far as I know, '水' is picked by @fantasai while working on writing modes, not a random Googler.

**Comment by sunhaitao** [comment]
@faceless2
You are right. Given that 'ch' only needs to be an approximation, any CJK characters in a font should be representative enough. There may be fonts that do not give each CJK character exactly the same measurement, but it is safe to assume that their sizes do not vary too much.

In that case, maybe we could just use the measurement of the first CJK character in the current font to avoid any confusion or download.

Using the measurement of ideographic space (全角空格, U+0300) is also a good choice. East Asians use it to arrange characters for quite a while.

Of course, just keep using the current work is also fine. It works, and it is real.

**Comment by jtyjty99999** [comment]
i prefer '永'

**Comment by jtyjty99999** [comment]
i prefer '永' instead of '水'. "Yongzi eight methods" is the rule of Chinese Calligraphy.

**Comment by tiansh** [comment]
> Speaking of font coverage, they are not the same in _web_ fonts. I don't know how things are going on in #3135 , but if `ic` requires implementations to download the font, then 永 (U+6C38) has higher chance to cause extra font download compared to 水 (U+6C34):

Web fonts for special using (brand name for example) won't include either one regardless what is chosen. But as 永 is a common used character, any web fonts for general using could easily include it if they want. And whenever the specification switched into 永 and people get use it, web fonts creators would/could/should change their fonts by including 永 in the most common "bucket". So that wont be a trouble here, IMO.

**Comment by chrishtr** [comment]
I discussed this issue with @xiaochengh, @kojiishi, @wangxianzhu and one other expert. We concluded that there should be no significant compat or interop risk with changing the spec text to use 永 instead of 水, since in basically all CJK fonts the characters have exactly the same width and height.

Given that, and since 永 makes much more sense to Chinese speakers (including those I consulted), I recommend changing the spec to use that character.

Note: In the Chromium implementation, we plan to still use 水 for the moment, because of issues like those mentioned in [this comment](https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1210082393). But again, that is just an implementation detail and will have no difference in the dimensions observed by web developers, and should *not* be how it's described in the spec or developer documentation. (I just mention it because otherwise there might be a performance concern with the change. We could change the implementation easily in the future if font buckets change.)

**Comment by nhnhwsnh** [comment]
I think `永`  is better.

**Comment by cdll** [comment]
`水` is not bad, but `永` is better. 🎉

**Comment by fantasai** [comment]
@chrishtr I would rather not have the spec and the implementations randomly diverge, so we should only change the spec if implementations are also committed to change. This would not be a difficult change in the implementations, actually: most of the work is in updating the W3C Recommendation and writing proper tests for it, which @yisibl has volunteered to do. (Note that the tests need to be subtle enough to account properly for non-square CJK fonts and for proportional fonts, which some designers are starting to experiment with.) Updating a W3C Recommendation involves a lot of tracking across time as various requirements (testing, review, implementation) are fulfilled, so if @yisibl is willing to follow through on all stages of the spec update, that will help a lot.

But again, the entire point of having specs is to define implementations, and in fact the W3C Recommendation process requires demonstration of conforming implementations, so unless we're willing to update the implementations we cannot update the specs either.

<hr>

I appreciate @myakura's analysis in https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1210082393 ; as @foolip notes in https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1210495600 the choice of 水 over 永 was intentional due to the frequency of usage, some practical effects of which we are seeing in the way font files are broken down.

There's also some subtle effects in the way characters are drawn and measured for optical alignment that probably make 水 slightly better than 永 in cases where we might need to measure the actual glyph ink area as a fallback for the `ICFT` baseline for fonts that don't have one specified. See @yisibl’s [note](https://github.com/w3c/csswg-drafts/issues/7577#issuecomment-1210493671) about CSS Inline Layout L3.

@gongpeione wrt “using 永 as the ic unit makes more sense in Chinese and also looks more professional”: You will never actually see this character, except in the specifications. And unless you have a font where each character is drawn in a different size, you will not know which character is used in the implementations. We only measuring its width and height to find the size of a typical CJK character.

@sunhaitao Appreciate your comments thinking through the issues with each character. The reason we're not using U+3000 IDEOGRAPHIC SPACE is because in some proportional fonts, it does not match the full width of a Han character. Most CJK fonts are not proportional, and so measuring any character is the same; but some few are not, and we need to accommodate them properly. This is why we can't use just any character.

@CoelacanthusHex They are the same in practice in most cases, but we have to gracefully handle proportional fonts and degenerate situations such as incomplete font metrics... 

<hr>

I've rephrased the definition of the `ic` unit in the specs to emphasize the conceptual definition over the implementation method, it now reads:

> **ic unit**
> Represents the typical _advance measure_ of CJK letters, and measured as the used _advance measure_ of the “水” (CJK water ideograph, U+6C34) glyph found in the font used to render it.

It was never the intention that the definition of _which_ CJK character we measure would be in user-facing documentation, as the main idea is that we are providing a measure that approximates the measure of a _typical_ CJK character. In most fonts, this should be equal to the measure of _any_ CJK character.

<hr>

Anyone commenting on this issue: **there are etiquette guidelines to commenting in bug reporting systems, the first and foremost of which is, don't spam the system with duplicate reports or comments**. In order for groups to work effectively in public, bug reporting systems need to _efficiently_ collect and represent relevant information. If you want to +1 this issue, add a thumbs-up to the existing comment. It is only worth commenting if you have additional relevant information or analysis to add, or if you have a question that is not already answered here.

(Also, don't insult people. This is also against etiquette.)

I want to emphasize that our goal in making decisions here is to **make things work as well as we can, in as many cases as we can**. Each decision we make is for some functional reason. We should correct mistakes, but keep in mind: we don't want the specs to be beautiful, we want the web pages that browsers render when following them to be beautiful.
